### PR TITLE
MM-18517: Plugin Marketplace (Phase 2)

### PR DESCRIPTION
**Pull request by lieut-data**
#### Summary
Introduces support for upgrading plugins, including prepackaged plugins, by preventing prepackaged plugins from installing on top of newer managed or unmanaged plugins. A few minor changes to support the webapp and e2e tests are included.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-18517

### CODE REVIEW

File: app/plugin.go [hunk 1]
```diff
@@ -164,10 +164,17 @@ func (a *App) InitPlugins(pluginDir, webappPluginDir string) {
 				return nil
 			}
 
-			if fileReader, err := os.Open(walkPath); err != nil {
+			fileReader, err := os.Open(walkPath)
```
Comment (streamer45) [inline app/plugin.go]: Any reason for not using `app.FileReader()` here? 
Also we should make sure we close it before returning.

File: app/plugin_install.go [hunk 2]
```diff
@@ -169,17 +182,44 @@ func (a *App) installPluginLocally(pluginFile io.ReadSeeker, replace bool) (*mod
 		return nil, model.NewAppError("installPluginLocally", "app.plugin.install.app_error", nil, err.Error(), http.StatusInternalServerError)
 	}
 
-	// Check that there is no plugin with the same ID
+	// Check for plugins installed with the same ID.
 	for _, bundle := range bundles {
-		if bundle.Manifest != nil && bundle.Manifest.Id == manifest.Id {
-			if !replace {
-				return nil, model.NewAppError("installPluginLocally", "app.plugin.install_id.app_error", nil, "", http.StatusBadRequest)
+		if bundle.Manifest != nil && bundle.Manifest.Id != manifest.Id {
```
Comment (crspeller) [inline app/plugin_install.go]: What if the bundle.Manifest is nil?

File: app/plugin_install.go [hunk 3]
```diff
@@ -169,17 +182,44 @@ func (a *App) installPluginLocally(pluginFile io.ReadSeeker, replace bool) (*mod
...
 			}
 		}
+
+		// Otherwise remove the existing installation prior to install below.
+		mlog.Debug("Removing existing installation of plugin before local install", mlog.String("plugin_id", bundle.Manifest.Id), mlog.String("version", bundle.Manifest.Version))
+		if err := a.removePluginLocally(manifest.Id); err != nil {
+			return nil, model.NewAppError("installPluginLocally", "app.plugin.install_id_failed_remove.app_error", nil, "", http.StatusBadRequest)
+		}
+
+		break
```
Comment (crspeller) [inline app/plugin_install.go]: The control flow here is a bit hard to follow. Maybe have the loop set some vars then continue top level?

File: app/plugin_install.go (hunk 2 above)
Comment (lieut-data) [reply app/plugin_install.go]: Good catch: this is cleaned up below.

File: app/plugin_install.go (hunk 3 above)
Comment (lieut-data) [reply app/plugin_install.go]: Agreed, and good observation. I've extracted an `existingManifest` check that clarifies this.
